pt-lua.c

```diff
@@ -0,0 +1,966 @@
+  int mwhite, mblack;
+  countframes(fnstack, &mwhite, &mblack);
+  /* Max levels of Lua stack. */
+  int mlevel = countlevels(L);
+
+  /* Total frames we are going to print. */
+  /* Black frames are used for switching and we will start from
+     Lua stack level 1. */
+  int nframes = mlevel + mwhite - mblack - 1;
+  /* Amount of frames printed. */
+  int pframes = 0;
+  /* Should we print ellipsis? */
+  bool ellipsis = nframes > (PT_LUA_TRACEBACK_TOP_THRESHOLD
+    + PT_LUA_TRACEBACK_BOTTOM_THRESHOLD);
+
+  luaL_Buffer buf;
+  luaL_buffinit(L, &buf);
+  lua_pushfstring(L, "%s\nstack traceback:", msg);
+  luaL_addvalue(&buf);
+
+  lua_Debug ar;
+  int level = 1;
+  const char *tname;
+
+  while(lua_getstack(L, level++, &ar)) {
+    /* Get information regarding the frame: name, source, linenumbers etc. */
+    lua_getinfo(L, "Slnf", &ar);
+
+    /* If the frame is a C frame. */
+    if(lua_iscfunction(L, -1)) {
+      if(index >= 0) {
+        /* Check whether this frame is tracked (C interface frames). */
+        int check = index;
+        while(stack[check].type != PALLENE_TRACER_FRAME_TYPE_LUA)
+          check--;
+
+        /* If the frame matches, we switch to printing Pallene frames. */
+        if(lua_tocfunction(L, -1) == stack[check].shared.c_fnptr) {
+          /* Now print all the frames in Pallene stack. */
+          for(; index > check; index--) {
+            lua_pushfstring(L, "\n    %s:%d: in function '%s'",
+              stack[index].shared.details->filename,
+              stack[index].line, stack[index].shared.details->fn_name);
+            MAKEITCOUNT();
+          }
+
+          /* 'check' idx is guaranteed to be a Lua interface frame.
+             Which is basically our 'stack' index at this point. So,
+             we simply ignore the Lua interface frame. */
+          index--;
+
+          /* We are done. */
+          continue;
+        }
+      }
+
+      /* Then it's an untracked C frame. */
+      if(pushglobalfuncname(L)) {
+        tname = lua_tostring(L, -1);
+        lua_pop(L, 1);
+      } else tname = "<?>";
+
+      lua_pushfstring(L, "\n    C: in function '%s'", tname);
+      MAKEITCOUNT();
+    } else {
+      /* It's a Lua frame. */
+
+      /* Do we have a name? */
+      if(*ar.namewhat != '\0') {
+        lua_pushfstring(L, "function '%s'", ar.name);
+        tname = lua_tostring(L, -1);
+        lua_pop(L, 1);
+      }
+      /* Is it the main chunk? */
+      else if(*ar.what == 'm')
+        tname = "<main>";
+      /* Can we deduce the name from the global table? */
+      else if(pushglobalfuncname(L)) {
+        lua_pushfstring(L, "function '%s'", lua_tostring(L, -1));
+        tname = lua_tostring(L, -1);
+        lua_pop(L, 2);
+      } else tname = "function '<?>'";
+
+      lua_pushfstring(L, "\n    %s:%d: in %s", ar.short_src,
+        ar.currentline, tname);
+      MAKEITCOUNT();
+    }
+  }
+
+  luaL_pushresult(&buf);
+  return 1;
+}
+#undef MAKEITCOUNT
+
```

Comment:
Please add a big comment here, explainig that everything below is copy pasted from lua.c (except for the xpcall that you changed). Don't forget to add the appropriate author & copyright information.